### ISSUE TITLE
Gathering typesystem infromation can be optimized

### ISSUE THREAD

**Issue by Therzok** [mono/monodevelop#6244]
Quite a few things here:
* Improve ProjectFile.get_ProjectVirtualPath efficiency (it's always computed, slow)
* Improve ProjectFile.set_Project on ProjectFile when it's done in its initializer
* Move ProjectSystemHandler.updatingProjectDataLock might be overzealous? Make it smaller scope, just projectDataMap manipulation?

```
0.90%   MoveNext  •  14,728 ms  •  221 calls  •  MonoDevelop.Projects.MSBuild.RemoteProjectBuilder+<SendRun>d__15.MoveNext
  0.90%   SetResult  •  14,728 ms  •  221 calls  •  System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetResult(TResult)
    0.90%   MoveNext  •  14,726 ms  •  221 calls  •  MonoDevelop.Projects.MSBuild.RemoteProjectBuilder+<Run>d__10.MoveNext
      0.90%   SetResult  •  14,725 ms  •  221 calls  •  System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetResult(TResult)
        0.90%   MoveNext  •  14,722 ms  •  221 calls  •  MonoDevelop.Projects.Project+<>c__DisplayClass150_0+<<RunMSBuildTarget>b__0>d.MoveNext
          0.90%   SetResult  •  14,715 ms  •  221 calls  •  System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetResult
            0.90%   MoveNext  •  14,711 ms  •  221 calls  •  MonoDevelop.Projects.Project+<RunMSBuildTarget>d__150.MoveNext
              0.85%   SetResult  •  13,975 ms  •  221 calls  •  System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetResult(TResult)
                0.85%   MoveNext  •  13,972 ms  •  221 calls  •  MonoDevelop.Projects.Project+<DoRunTarget>d__149.MoveNext
                  0.85%   SetResult  •  13,971 ms  •  221 calls  •  System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetResult(TResult)
                    0.66%   MoveNext  •  10,829 ms  •  220 calls  •  MonoDevelop.Projects.Project+CachingCoreCompileEvaluator+<GetItemsFromCoreCompileDependenciesAsync>d__5.MoveNext
                      0.54%   SetResult  •  8,813 ms  •  220 calls  •  System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetResult(TResult)
                        0.54%   MoveNext  •  8,811 ms  •  110 calls  •  MonoDevelop.Projects.Project+<OnGetSourceFiles>d__78.MoveNext
                          0.33%   SetResult  •  5,420 ms  •  110 calls  •  System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetResult(TResult)
                            0.31%   MoveNext  •  5,134 ms  •  109 calls  •  MonoDevelop.Ide.TypeSystem.MonoDevelopWorkspace+ProjectSystemHandler+<LoadProject>d__14.MoveNext
                               0.18%   Enter  •  2,991 ms  •  109 calls  •  System.Threading.Monitor.Enter(Object, ref Boolean)
                              0.10%   CreateDocuments  •  1,621 ms  •  109 calls  •  MonoDevelop.Ide.TypeSystem.MonoDevelopWorkspace+ProjectSystemHandler.CreateDocuments(ProjectData, Project, CancellationToken, ProjectFile[], ProjectData)
                                0.06%   CreateDocumentInfo  •  987 ms  •  7,110 calls  •  MonoDevelop.Ide.TypeSystem.MonoDevelopWorkspace+ProjectSystemHandler.CreateDocumentInfo(SolutionData, String, ProjectData, ProjectFile)
                                  0.05%   GetFolders  •  748 ms  •  7,110 calls  •  MonoDevelop.Ide.TypeSystem.MonoDevelopWorkspace+ProjectSystemHandler.GetFolders(String, ProjectFile)
                                    ►0.04%   get_ProjectVirtualPath  •  699 ms  •  7,110 calls  •  MonoDevelop.Projects.ProjectFile.get_ProjectVirtualPath
                                     0.00%   4 functions hidden  •  46 ms total  •  28440 calls total
                                   0.01%   7 functions hidden  •  131 ms total  •  71100 calls total
                                  ►0.01%   ResolveLinks  •  101 ms  •  7,110 calls  •  MonoDevelop.Core.FilePath.ResolveLinks
                                0.01%   CanGenerateAnalysisContextForNonCompileable  •  178 ms  •  4,621 calls  •  MonoDevelop.Ide.TypeSystem.MonoDevelopWorkspace+ProjectSystemHandler.CanGenerateAnalysisContextForNonCompileable(Project, ProjectFile)
                                  ►0.01%   GetMimeTypeForUri  •  89 ms  •  4,621 calls  •  MonoDevelop.Ide.DesktopService.GetMimeTypeForUri(String)
                                  ►0.01%   GetTypeSystemParserNode  •  86 ms  •  4,621 calls  •  MonoDevelop.Ide.TypeSystem.TypeSystemService.GetTypeSystemParserNode(String, String)
                                   0.00%   3 functions hidden  •  1 ms total  •  13863 calls total
                                0.01%   MoveNext  •  178 ms  •  4,621 calls  •  MonoDevelop.Ide.TypeSystem.MonoDevelopWorkspace+ProjectSystemHandler+<GenerateProjections>d__22.MoveNext
                                  ►0.01%   GetTypeSystemParserNode  •  88 ms  •  4,621 calls  •  MonoDevelop.Ide.TypeSystem.TypeSystemService.GetTypeSystemParserNode(String, String)
                                  ►0.01%   GetMimeTypeForUri  •  88 ms  •  4,621 calls  •  MonoDevelop.Ide.DesktopService.GetMimeTypeForUri(String)
                                   0.00%   3 functions hidden  •  1 ms total  •  13863 calls total
                                ►0.01%   ResolveLinks  •  122 ms  •  7,110 calls  •  MonoDevelop.Core.FilePath.ResolveLinks
                                ►0.01%   IsCompileable  •  83 ms  •  11,731 calls  •  MonoDevelop.Projects.Project.IsCompileable(String)
                                 0.00%   22 functions hidden  •  60 ms total  •  123269 calls total
                              ►0.03%   CreateCompilationOptions  •  455 ms  •  108 calls  •  MonoDevelop.CSharp.Project.CSharpCompilerParameters.CreateCompilationOptions
                               0.00%   27 functions hidden  •  60 ms total  •  3160 calls total
                            ►0.02%   MoveNext  •  284 ms  •  1 call  •  MonoDevelop.DotNetCore.DotNetCoreProjectExtension+<OnGetSourceFiles>d__44.MoveNext
                          ►0.21%   ToList  •  3,390 ms  •  110 calls  •  System.Linq.Enumerable.ToList(IEnumerable)
                           0.00%   5 functions hidden  •  0 ms total  •  550 calls total
                      0.12%   ProcessMSBuildItems  •  2,015 ms  •  110 calls  •  MonoDevelop.Projects.Project+CachingCoreCompileEvaluator.ProcessMSBuildItems(IEnumerable, Project)
                        0.10%   set_Project  •  1,571 ms  •  7,596 calls  •  MonoDevelop.Projects.ProjectItem.set_Project(Project)
                          0.10%   OnProjectSet  •  1,569 ms  •  7,596 calls  •  MonoDevelop.Projects.ProjectFile.OnProjectSet
                            ►0.05%   get_Include  •  783 ms  •  7,596 calls  •  MonoDevelop.Projects.ProjectFile.get_Include
                            ►0.05%   get_ProjectVirtualPath  •  779 ms  •  7,596 calls  •  MonoDevelop.Projects.ProjectFile.get_ProjectVirtualPath
                             0.00%   4 functions hidden  •  3 ms total  •  30384 calls total
                        ►0.01%   FromMSBuildPath  •  217 ms  •  7,596 calls  •  MonoDevelop.Projects.MSBuild.MSBuildProjectService.FromMSBuildPath(String, String)
                        ►0.01%   ProjectFile..ctor  •  157 ms  •  7,596 calls  •  MonoDevelop.Projects.ProjectFile..ctor(String, String)
                         0.00%   14 functions hidden  •  61 ms total  •  69244 calls total
                       0.00%   11 functions hidden  •  3 ms total  •  1760 calls total
                    ►0.19%   MoveNext  •  3,139 ms  •  111 calls  •  MonoDevelop.Projects.DotNetProject+<RunResolveAssemblyReferencesTarget>d__111.MoveNext
                   0.00%   3 functions hidden  •  1 ms total  •  663 calls total
              ►0.01%   MSBuildPropertyEvaluated..ctor  •  225 ms  •  173,504 calls  •  MonoDevelop.Projects.MSBuild.MSBuildPropertyEvaluated..ctor(MSBuildProject, String, String, String, Boolean)
               0.01%   52 functions hidden  •  201 ms total  •  1199763 calls total
               0.01%   set_Item  •  178 ms  •  173,504 calls  •  System.Collections.Generic.Dictionary`2.set_Item(TKey, TValue)
           0.00%   3 functions hidden  •  7 ms total  •  663 calls total
       0.00%   4 functions hidden  •  0 ms total  •  884 calls total
   0.00%   2 functions hidden  •  0 ms total  •  442 calls total
```

> VS bug [#700297](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/700297)

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-39] Comment by **xamarinc**



Fixed in version 8.4.0.164 (master)



Pull Request #7901 merged by: Lluis Sanchez  
 Author: mono  
 Commit: a8cc97b50f1fece90690febd43761c157e773fe8 (mono/monodevelop)

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-42] Comment by **xamarinc**



Fixed in version 8.2.0.948 (release-8.2)



Pull Request #8042 merged by: Lluis Sanchez  
 Author: mono  
 Commit: 9464f06e56a180a7c0b6c15ebbf684790f3421a3 (mono/monodevelop)

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-44] Comment by **xamarinc**



Fixed in version 8.2.0.970 (release-8.2)



Pull Request #8042 merged by: Lluis Sanchez  
 Author: mono  
 Commit: 9464f06e56a180a7c0b6c15ebbf684790f3421a3 (mono/monodevelop)

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-45] Comment by **xamarinc**



Fixed in version 8.4.0.203 (release-8.3)



Pull Request #7901 merged by: Lluis Sanchez  
 Author: mono  
 Commit: a8cc97b50f1fece90690febd43761c157e773fe8 (mono/monodevelop)

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-46] Comment by **xamarinc**



Fixed in version 8.3.0.578 (release-8.3)



Pull Request #7901 merged by: Lluis Sanchez  
 Author: mono  
 Commit: a8cc97b50f1fece90690febd43761c157e773fe8 (mono/monodevelop)

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-50] Comment by **xamarinc**



Fixed in version 8.6.0.802 (master)



Pull Request #7901 merged by: Lluis Sanchez  
 Author: mono  
 Commit: a8cc97b50f1fece90690febd43761c157e773fe8 (xamarin/monodevelop)
